google-ads-examples/src/main/java/com/google/ads/googleads/examples/shoppingads/GetProductBiddingCategoryConstant.java

```diff
@@ -0,0 +1,193 @@
+    SearchGoogleAdsRequest request =
+        SearchGoogleAdsRequest.newBuilder()
+            .setPageSize(PAGE_SIZE)
+            .setCustomerId(Long.toString(customerId))
+            .setQuery(query)
+            .build();
+
+    // Creates the Google Ads Service Client.
+    try (GoogleAdsServiceClient googleAdsServiceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+
+      // Creates a list of top level category nodes.
+      List<CategoryNode> rootCategories = new ArrayList<>();
+      // Creates a map of category ID to category node for all categories found in the results.
+      HashMap<String, CategoryNode> biddingCategories = new HashMap<>();
+
+      // Performs the search request.
+      SearchPagedResponse response = googleAdsServiceClient.search(request);
+      for (GoogleAdsRow googleAdsRow : response.iterateAll()) {
+        ProductBiddingCategoryConstant productBiddingCategory =
+            googleAdsRow.getProductBiddingCategoryConstant();
+
+        String name = productBiddingCategory.getLocalizedName().getValue();
+        String resourceName = productBiddingCategory.getResourceName();
+        CategoryNode node = biddingCategories.get(resourceName);
+        if (node == null) {
+          node = new CategoryNode(resourceName, name);
+          biddingCategories.put(resourceName, node);
+        } else if (node.name == null) {
+          // Ensures that the name attribute for the node is set. Name will be null for nodes added
+          // to biddingCategories as a result of being a parentNode below.
+          node.name = name;
+        }
+
+        if (productBiddingCategory.hasProductBiddingCategoryConstantParent()) {
+          String parentResourceName =
+              productBiddingCategory.getProductBiddingCategoryConstantParent().getValue();
+          CategoryNode parentNode = biddingCategories.get(parentResourceName);
+          if (parentNode == null) {
+            parentNode = new CategoryNode(parentResourceName);
+            biddingCategories.put(parentResourceName, parentNode);
+          }
+          parentNode.children.add(node);
+        } else {
+          rootCategories.add(node);
+        }
+      }
+      displayCategories(rootCategories, "");
+    }
+  }
+
+  /**
+   * Recursively prints out each category node and its children.
+   *
+   * @param categories the categories to print.
+   * @param prefix the string to print at the beginning of each line of output.
+   */
+  private static void displayCategories(List<CategoryNode> categories, String prefix) {
+    for (CategoryNode category : categories) {
+      System.out.printf("%s%s [%s]%n", prefix, category.name, category.resourceName);
+      displayCategories(category.children, String.format("%s%s > ", prefix, category.name));
+    }
+  }
+
+  /** Node that tracks a product bidding category's id, name, and child nodes. */
+  private static class CategoryNode {
+    final String resourceName;
```

Comment:
private final for all fields and use getters.

Reply:
+1 on `final` and getters.
`name` has to be nullable since the example adds each parent under its resource name even if it first discovers that parent via one of its children.